### DISCUSSION TITLE
Request: llvm: add flang Fortran compiler

### DISCUSSION THREAD

**Discussion by scivision** [Homebrew/discussions#5631] [Everyday usage] [answered]
### Output of `brew config`

```shell
N/A
```

### Output of `brew doctor`

```shell
N/A
```

### Description of issue

Flang the LLVM Fortran compiler has matured in LLVM 19 to where Flang can now compile most Fortran code, even for major projects. I would like to have "flang" be installed as part of "llvm" or at least make "flang" its own formula.

What I've been doing is to build Flang with a [script](https://gist.github.com/scivision/33bd9e17c9520d07be0448fe61541605), in case this is helpful to see what options are used.

[reference, flang has progressed beyond this](https://github.com/orgs/Homebrew/discussions/2753)N

**Comment by alebcay** [comment]
Some related work is currently happening in https://github.com/Homebrew/homebrew-core/pull/191829.

**Comment by mathomp4** [comment]
So, I saw today that `flang` is now in Brew and decided to try it out. Unfortunately, I think I'm missing something:

```
❯ which flang-newv
/Users/mathomp4/.homebrew/brew/bin/flang-new
❯ cat test.F90
program main

end
❯ flang-new test.F90
Undefined symbols for architecture arm64:
  "std::exception::~exception()", referenced from:
      std::bad_variant_access::~bad_variant_access() in libFortranRuntime.a[19](edit-output.cpp.o)
  "std::terminate()", referenced from:
      ___clang_call_terminate in libFortranRuntime.a[5](assign.cpp.o)
  "typeinfo for std::bad_variant_access", referenced from:
      std::__1::__throw_bad_variant_access[abi:ue170006]() in libFortranRuntime.a[19](edit-output.cpp.o)
  "vtable for std::bad_variant_access", referenced from:
      std::__1::__throw_bad_variant_access[abi:ue170006]() in libFortranRuntime.a[19](edit-output.cpp.o)
   NOTE: a missing vtable usually means the first non-inline virtual member function has no definition.
  "___cxa_allocate_exception", referenced from:
      std::__1::__throw_bad_variant_access[abi:ue170006]() in libFortranRuntime.a[19](edit-output.cpp.o)
  "___cxa_begin_catch", referenced from:
      ___clang_call_terminate in libFortranRuntime.a[5](assign.cpp.o)
  "___cxa_throw", referenced from:
      std::__1::__throw_bad_variant_access[abi:ue170006]() in libFortranRuntime.a[19](edit-output.cpp.o)
  "___gxx_personality_v0", referenced from:
      /Users/mathomp4/.homebrew/brew/Cellar/flang/19.1.0/libexec/lib/libFortranRuntime.a[52](stop.cpp.o)
ld: symbol(s) not found for architecture arm64
flang-new: error: linker command failed with exit code 1 (use -v to see invocation)
```

I've tried this with both `CC` as Apple clang and with `CC` as brew llvm clang. Staring at the `brew info llvm` output I also tried:

```
flang-new -L/Users/mathomp4/.homebrew/brew/opt/llvm/lib -I/Users/mathomp4/.homebrew/brew/opt/llvm/include test.F90

flang-new -L/Users/mathomp4/.homebrew/brew/opt/llvm/lib/c++ -L/Users/mathomp4/.homebrew/brew/opt/llvm/lib/unwind -lunwind test.F90

flang-new -L/Users/mathomp4/.homebrew/brew/opt/llvm/lib/unwind -lunwind test.F90
```

But they all fail. So I'm obviously missing something.

NOTE: One issue could be that on this machine I'm not allowed to install to `/opt`. But that doesn't stop brew llvm clang from working:
```
❯ /Users/mathomp4/.homebrew/brew/opt/llvm/bin/clang test.c
❯ ./a.out
Hello, World!%
```

**Reply by carlocab** [reply]
Try
```
flang-new test.F90 -lc++
```

**Reply by mathomp4** [reply]
Ooh. That does work:
```
❯ flang-new test.F90 -lc++
❯ echo $?
0
```

So. Huh. I guess that's something I can't really code in my modulefile. Time to append to a lot configures and CMake.

**Reply by scivision** [reply]
Is this `-lc++` something that the Flang Homebrew package should automatically do as part of its self-defined toolchain?

I.e. like GCC automatically adds `-lgfortran` when linking a C or C++ main program with a Fortran library.

I experience the same symptoms as @mathomp4 

For now what I do is add to my source file ~/flang.sh (say)

```sh
export FC=flang-new 

export LDFLAGS=-lc++
```

The LDFLAGS is consumed by build systems such as CMake--be sure to make a fresh project build directory as prior attempts to use the compiler will mess up the detection of the compiler by CMake.

**Reply by mathomp4** [reply]
Interesting. That does not work for me:
```
❯ export LDFLAGS='-lc++'
❯ flang-new test.F90
Undefined symbols for architecture arm64:
  "std::exception::~exception()", referenced from:
      std::bad_variant_access::~bad_variant_access() in libFortranRuntime.a[19](edit-output.cpp.o)
...
```

**Reply by scivision** [reply]
Right, you would need the LDFLAGS to be specified on the command line if manually compiling a single file.

LDFLAGS is intended to be consumed by build systems like Meson or CMake on their first configure step--make a fresh build directory to clear out prior unsuccessful compiler detection attempts.

I have to do things like this for other compilers too--but not for GCC. It would be nice if Flang Homebrew package would "know" it needs "-lc++" implicit in every link command by `flang-new`.

**Reply by mathomp4** [reply]
Ahh. Feelin' dumb now. I guess I thought you were applying that to all.

**Reply by mathomp4** [reply]
Meanwhile, my home laptop (which can use the bottle) seems to have a different error:

```
❯ which flang-new
/opt/homebrew/bin/flang-new
❯ flang-new test.F90
ld: library 'System' not found
flang-new: error: linker command failed with exit code 1 (use -v to see invocation)
❯ flang-new test.F90 -lc++
ld: library 'c++' not found
flang-new: error: linker command failed with exit code 1 (use -v to see invocation)
```

Huh.

**Reply by scivision** [reply]
appending the `-v` flag to the flang-new command will show the command line used.

It sounds like you need the LIBRARY_PATH set. Try something like:

```sh
export LIBRARY_PATH=/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/lib
# where libSystem.tbd is
# where libc++.tbd is
```

You can also just put that on the command line like

```sh
flang-new main.f90 -L/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/lib -lc++
```

These are flags that the Homebrew devs may want to make part of their Flang package, i.e. injecting them into the final command line emitted by user `flang-new` commands.

**Reply by cho-m** [reply]
In theory, it shouldn't be needed as Flang doesn't support exception handling.

However, the `-DLLVM_ENABLE_EH=OFF` when building Flang is probably just a no-op (specifically a [CMake error bypass](https://github.com/llvm/llvm-project/blob/main/flang/CMakeLists.txt#L17-L23)) when linking to an existing LLVM built with `-DLLVM_ENABLE_EH=ON`. I would guess the option is actually set by `LLVMConfig.cmake` which has:
```cmake
set(LLVM_ENABLE_EH ON)
```

May also be worth checking if CMake `LLVMFlang` (and maybe other build systems) work as is, which would provide information on how build systems think this situation should be handled.

Related: https://github.com/llvm/llvm-project/issues/59353

**Reply by scivision** [reply]
CMake isn't able to configure a trivial Fortran project with Homebrew Flang unless 

```sh
LDFLAGS=-lc++ cmake -Bbuild
```
due to similar errors when detecting the Flang compiler

```cmake
project(hello LANGUAGES Fortran)
```

Fails without the above LDFLAGS

**Reply by cho-m** [reply]
We could consider switching to default of `LLVM_ENABLE_EH=OFF` if that is what most distros/repositories do.

This will impact users who want to make use of C++ exceptions propagating through LLVM code, but I think majority of major repositories don't enable it anyways. Verified this for:
* Arch Linux
* Debian
* Fedora (aarch64 rpm)
* MacPorts
* NixOS
* [Spack when built with `+flang`](https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/llvm/package.py#L846-L854)

EH is enabled for:
* Fedora (x86_64 rpm)
* [Gentoo](https://gitweb.gentoo.org/repo/gentoo.git/tree/sys-devel/llvm/llvm-19.1.1.ebuild#n397) - I don't think they support Flang

**Reply by carlocab** [reply]
> Is this `-lc++` something that the Flang Homebrew package should automatically do as part of its self-defined toolchain?

This should be handled in https://github.com/Homebrew/homebrew-core/pull/192611, which should hopefully merge some time this week.

**Reply by carlocab** [reply, marked as answer]
Homebrew/homebrew-core#192611 has been merged. You may need to `brew update && brew reinstall flang` to pick up the fix.

**Reply by mathomp4** [reply]
Ooh! First:
```
==> Caveats
Homebrew LLVM is built with LLVM_ENABLE_EH=ON, so binaries built by `flang-new`
require linkage to the C++ standard library. `flang-new` is configured to do this
automatically.
```
Nice.

And on my home mac:
```
❯ flang-new test.F90
❯ echo $?
0
```

So it works okay in the default `/opt/homebrew` area. I'll check my work laptop (where it's a non-standard install) on Monday.

**Reply by scivision** [reply]
Yes, #192611 fixes Flang for me. I no longer have to set LDFLAGS at all.

All I do now to use Flang is:

```sh
export FC=flang-new
```

Flang works to link Fortran-C-C++ binaries with AppleClang, Clang, and GCC.
